DOM: Allow copying text from non-text input elements

mcsf opened the pull request:
## What?

Fixes #32148.

This pull request guarantees that a user is able to copy the contents of any `<input>` elements inside blocks using their browser's native actions (<kbd>Ctrl-C</kbd>, <kbd>Command-C</kbd>).

<img width="383" alt="Captura de ecrã 2022-04-08, às 16 10 29" src="https://user-images.githubusercontent.com/150562/162473526-0ed9fff9-351a-44b4-ab27-9f619ecf7cab.png">

## Why?

This pull request fixes a bug in which certain types of input were not copiable — specifically, any "non-text" input, such as `number` or even `email`. Typically, when this bug occurred, the block editor would ignore the browser's native copying behaviour, thus resulting in the copying of the entire block — instead of just the selected content in the input element.

## How?

The selection state of non-text input elements is opaque to Gutenberg, [per the HTML spec](https://html.spec.whatwg.org/multipage/input.html#do-not-apply). Unaware of this
nuance, `inputFieldHasUncollapsedSelection( element )` would incorrectly report that some non-text input element had no text selected when in fact it did. This caused the block editor to take over `copy` events to copy whole blocks, preventing the user from copying what they had actually selected inside the input element in question.

**The trade-off in this commit** is to always assume that there could be an uncollapsed selection in an active element if that element's selection state is opaque.

### Caveat

The trade-off means that, if the user places the caret inside a non-text input element but has no text selected (according to the jargon, this means that the selection is _collapsed_), then pressing the copy shortcut will no longer let the block editor copy the whole block. Text inputs (`input="text"` or `"url"`) and rich-text elements should not be affected by this.

## Testing Instructions

For specifics, see instructions in #32148.

- **Fixed:** Ensure that the contents of non-text inputs can be copied using the native keyboard shortcut. Try this using `<TextControl type="number"` inside a block, as well as `type="email"`, `type="time"`.
- **Unchanged:** Ensure that the contents of text inputs can be copied using the native keyboard shortcut. Try this using `<TextControl type="text"` inside a block, as well as `type="url"`.
- **Unchanged:** Ensure that an entire block can be copied with the native keyboard shortcut when the caret is placed inside a text input with no selection.
- **Unchanged:** Ensure that copying is otherwise unaffected: the copying of rich-text content, of whole blocks, of multiple blocks, etc.

<details>
<summary><strong>Attached:</strong> The diff I applied to the Paragraph block for my own testing</summary>
<pre><code>diff --git a/packages/block-library/src/paragraph/edit.js b/packages/block-library/src/paragraph/edit.js
index c6a2988a3d..8795d868f8 100644
--- a/packages/block-library/src/paragraph/edit.js
+++ b/packages/block-library/src/paragraph/edit.js
@@ -7,10 +7,12 @@ import classnames from 'classnames';
  * WordPress dependencies
  */
 import { __, _x, isRTL } from '@wordpress/i18n';
+import { useState } from '@wordpress/element';
 import {
 	ToolbarDropdownMenu,
 	ToggleControl,
 	__experimentalToolsPanelItem as ToolsPanelItem,
+	TextControl,
 } from '@wordpress/components';
 import {
 	AlignmentControl,
@@ -56,6 +58,11 @@ function ParagraphBlock( {
 } ) {
 	const { align, content, direction, dropCap, placeholder } = attributes;
 	const isDropCapFeatureEnabled = useSetting( 'typography.dropCap' );
+
+	const [ fooText, setFooText ] = useState( 'hello' );
+	const [ fooEmail, setFooEmail ] = useState( 'foo@foo.bar' );
+	const [ fooNumber, setFooNumber ] = useState( 42 );
+
 	const blockProps = useBlockProps( {
 		className: classnames( {
 			'has-drop-cap': dropCap,
@@ -149,6 +156,21 @@ function ParagraphBlock( {
 				__unstableEmbedURLOnPaste
 				__unstableAllowPrefixTransformations
 			/>
+			&lt;TextControl
+				type="text"
+				value={ fooText }
+				onChange={ ( v ) => ( console.log( v ), setFooText( v ) ) }
+			/>
+			&lt;TextControl
+				type="email"
+				value={ fooEmail }
+				onChange={ ( v ) => ( console.log( v ), setFooEmail( v ) ) }
+			/>
+			&lt;TextControl
+				type="time"
+				value={ fooNumber }
+				onChange={ ( v ) => ( console.log( v ), setFooNumber( v ) ) }
+			/>
 		</>
 	);
 }
</code></pre>
</details>

mcsf committed:
DOM: Fix inputFieldHasUncollapsedSelection for non-text inputs

The selection state of non-text input elements (e.g. `number`, but even
`email`) is opaque to Gutenberg, per the HTML spec [1]. Unaware of this
nuance, `inputFieldHasUncollapsedSelection` would incorrectly report
that some non-text input element had no text selected when in fact it
did. This caused the block editor to take over `copy` events to copy
whole blocks, preventing the user from copying what they had actually
selected inside the input element in question.

The workaround in this commit is to always assume that there could be an
uncollapsed selection in an active element if that element's selection
state is opaque.

[1]: https://html.spec.whatwg.org/multipage/input.html#do-not-apply

mcsf commented on packages/dom/src/dom/is-text-field.js:
```diff
@@ -26,6 +26,7 @@ export default function isTextField( node ) {
 		'reset',
 		'submit',
 		'number',
+		'email',
```
Could any component by negatively affected by this spec correction?

mcsf replied:
Also, I really dislike `isTextField`'s exclusion approach. Since, in fact, _most_ input types are not text fields, we might be better off with a stricter inclusion approach, i.e. `[ 'text', ... ].includes( type )`.

ntsekouras replied:
I don't think anything would be negatively affected by this.. And I also think a reversed approach would be okay, but we would really need to make sure we have all the needed types 😄

mcsf committed:
Add input type 'time' to isTextField's blacklist

mcsf committed:
Unit tests: update expectations for isTextField()

ntsekouras commented on packages/dom/src/dom/input-field-has-uncollapsed-selection.js:
```diff
@@ -18,9 +18,25 @@ import isNumberInput from './is-number-input';
  */
 export default function inputFieldHasUncollapsedSelection( element ) {
-	if ( ! isTextField( element ) && ! isNumberInput( element ) ) {
+	if ( ! isTextField( element ) && ! isHTMLInputElement( element ) ) {
```
How about creating a new util here(instead of `isHTMLInputElement`) that wraps the checks for the specific inputs we want to let browsers handle it. I think these would be `number, email and time`? In that case we would add an extra same check below, if there is no selection but this is true.

Now if we have any other inputs like button, etc.. would return true and I don't think we need to do this.

mcsf replied:
I'm not sure I understood your question; could you show me?

ntsekouras replied:
In several places in the code base we use the `isNumberInput` for the exact same reason of this PR: trying also include `input[type="number"]` for some selection related actions/calculations. What I suggest is to make a new util that will include `isNumberInput` and similar checks as if we had `isEmailInput and isTimeInput`. Makes sense?

mcsf replied:
Thanks, I understand the point now. What's interesting is that, the more I look at this PR and at existing usage of `isNumberInput`, the more I'm confused. :) I'll comment in more detail in the general thread.

ntsekouras reviewed (commented):
Thanks for the PR Miguel!

I think this is the right direction, but we need to make sure we update if needed related code to this. For example in [documentHasSelection](https://github.com/WordPress/gutenberg/blob/trunk/packages/dom/src/dom/document-has-selection.js#L19) util we perform similar checks and we should probably take into account the added `email,time`. 

Also it's interesting that while we had code for checking `number` type, since it has no `selectionStart/End` we would always return false 😄

mcsf commented:
Part of what makes these components tricky is that browsers implement these interfaces differently. My notes in the following paragraph are sort of a recap, but I find it helpful even just for myself to explain it again:

The [spec states](https://html.spec.whatwg.org/#dom-textarea/input-selectionstart) that `selectionStart` and `selectionEnd` **only apply** to input types _text, search, url, tel, password_, and it states that, for any other types, the attributes `selectionStart` and `selectionEnd` **should be null**. Safari goes against both prescriptions: it extends selection attributes to input types **number and email**, while any attempt to read these attributes in other input types **throws a type error**.

Here's a [quick codesandbox script](https://codesandbox.io/s/amazing-mahavira-so62eh?file=/src/index.js) to try on different browsers. Red indicates input types deemed "opaque" — i.e., that don't support `selectionStart`, whether because it is `null` or because it throws an error. Green indicates a successful selection read. Notice that, in Safari, towards the bottom, types `number` and `email` show up in green, whereas in Firefox and Chrome only the first five types are green (following the spec).

I've been going back and forth on a question: since Safari is arguably being "useful" by supporting selection APIs in a wider range of input types, should we let users benefit from this via browser-dependent behaviour, or should we prioritise behaviour that is consistent across browsers and up to spec?

mcsf commented:
> In several places in the code base we use the `isNumberInput` for the exact same reason of this PR: trying also include `input[type="number"]` for some selection related actions/calculations. What I suggest is to make a new util that will include `isNumberInput` and similar checks as if we had `isEmailInput and isTimeInput`.

My other question, prompted by Nik's comment, is around `isNumberInput`. Luckily, it is only used twice in Gutenberg — in `documentHasSelection` and `inputFieldHasUncollapsedSelection` — but I wondered why number inputs receive special treatment in those situations. Reading the PR that introduced it, #21457, it seems that number inputs were special-cased simply because it was the only non-text input type where the issue was felt at the time. (As a side note, there is a fundamental bug in that function due to `!! node.valueAsNumber` returning false if the value is the number 0.)

This leads me to think that there is no reason to continue to treat numbers differently. Ultimately, all these selection-checking functions serve one purpose: to determine what to do when the user uses general keyboard shortcuts to copy, cut, and paste content.

Below are two tables describing what I think is the current **intended** behaviour in Gutenberg. For simplicity, ignore contentEditable elements.

**When pasting:**

focus \ action             | paste
-|-
text input                    | paste into field (= defer to default behaviour)
opaque input              | paste into field (= defer to default behaviour)
elsewhere in a block  | replace whole block with clipboard content (= prevent default, intervene)

Pasting is easier, because it doesn't care whether there is a collapsed or uncollapsed selection in the active element. Also note that, ideally, opaque inputs should behave just like text ones — browser permitting.

Copying and cutting is more complicated, due to the desire to distinguish collapsed and uncollapsed selections. In **bold** I highlight what this PR was initially aiming to fix (i.e. conform with the intended behaviour). In _italic_ I highlight what this PR was intentionally dismissing as a trade-off.

**When copying or cutting:**

focus \ action                        | copy                             | cut
-|-|-
text input (uncollapsed)       | copy input value         | cut input value
opaque input (uncollapsed) | **copy input value**  | **cut input value**
text input (collapsed)            | copy whole block       | cut whole block
opaque input (collapsed)      | _copy whole block_  | _cut whole block_
elsewhere in a block              | copy whole block      | copy whole block

## Solution one

Based on these tables, I'd like to drop usage of `isNumberInput` in our selection functions (although we can't remove the function, as it is public) and just operate based on whether the input is "opaque". This is a continuation of what this PR is already doing. The open question here is whether we should strictly adhere to the spec or whether we should go with what each browser allows.

## Solution two

Although solution 1 works well for me, I'm also tempted to simplify this drastically, removing the collapsed/uncollapsed distinction, which in turn eliminates the need to inspect the input:

focus \ action                        | copy                                                   | cut
-|-|-
input                                      | copy input value (browser default)  | cut input value (browser default)
(contentEditable)                 | (handled by RichText)                        | (handled by RichText)
elsewhere in a block            | copy whole block (WritingFlow)       | cut whole block (WritingFlow)

Thoughts? Preferences? /cc @ntsekouras, @youknowriad

mcsf committed:
inputFieldHasUncollapsedSelection: Rewrite comments, default to true

mcsf committed:
documentHasSelection: Accept any input types

... as oppposed to just text inputs (isTextField) and number inputs
(isNumberInput).

youknowriad commented:
@mcsf I don't feel like I understand all the implications but the reasoning makes sense and I'm always for simplifications. So I'm onboard with solution 2.

ntsekouras commented:
>(As a side note, there is a fundamental bug in that function due to !! node.valueAsNumber returning false if the value is the number 0.)

That's true 😄 . We should have a follow up on that one.

ntsekouras reviewed (approved):
The approach makes sense to me and code looks good. Thanks Miguel!

mcsf commented:
> @mcsf I don't feel like I understand all the implications but the reasoning makes sense and I'm always for simplifications. So I'm onboard with solution 2.

For the record, I paired with Nik today and we looked at both solutions, and we decided in favour of the first one. I had overestimated the simplification benefits of Solution 2, and I'd rather keep the special treatment of uncollapsed selections if that means that there's a higher likelihood that the editor will do what the user wanted.

Going to merge. :) Thanks, everyone.
